### ISSUE TITLE
Race condition in shutdown with ext/Event loop

### ISSUE THREAD

**Issue by brstgt** [binsoul/net-mqtt-client-react#10]
I am receiving a lot of E_WARNINGs as an already closed stream is closed during disconnect.

There is a race condition between active and passive connection closing. Normally the server closes the connection after a successful disconnect message. So the active connection close should not be necessary and the client should be informed by the stream about the closed connection. However, if sth goes wrong, there should be a fallback.

I suggest this
```
        $deferred = new Deferred();

        $resolved = false;
        $finishDisconnect = function($connection) use ($deferred, &$resolved) {
            if ($resolved) {
                return;
            }
            $resolved = true;
            $this->isDisconnecting = false;
            $this->emit('disconnect', [$connection, $this]);
            $deferred->resolve($connection);
        };

        $timer = null;
        $this->once('close', function($connection) use ($finishDisconnect, &$timer) {
            if ($timer) {
                $this->loop->cancelTimer($timer);
            }
            $finishDisconnect($connection);
        });

        $this->startFlow(new OutgoingDisconnectFlow($this->connection), true)
            ->then(function (Connection $connection) use ($finishDisconnect, &$timer) {
                $this->isConnected = false;

                $timer = $this->loop->addTimer(0.1, function () use ($connection, $finishDisconnect, &$timer) {
                    $timer = null;
                    if ($this->stream !== null) {
                        $finishDisconnect($connection);
                        $this->stream->close();
                    }
                });
            })
            ->otherwise(function () use ($deferred, &$resolved) {
                $this->isDisconnecting = false;
                if ($resolved) {
                    return;
                }
                $resolved = true;
                $deferred->reject($this->connection);
            });
```

**Comment by brstgt** [comment]
Isnt this project maintained any more?

**Comment by binsoul** [comment]
Better late than never... I released version 0.4.0 which waits for passive connection closing.
